src/CommandingOfficers/Tech.java

```diff
@@ -116,63 +119,78 @@ public int getRepairPower()
-    GameEventQueue events = super.initTurn(map);
-
-    return events;
+    long exclude = UnitModel.TROOP;
+    ArrayList<UnitModel> typesToOverCharge = commander.getAllModelsNot(exclude);
+    return typesToOverCharge;
   }
 
   /** Heal all units by the specified amount, allowing HP>10, and provide a buff. */
   private static class OverchargeAbility extends CommanderAbility
   {
-    private CODamageModifier damageBuff;
-    private CODefenseModifier defenseBuff;
+    private static final long serialVersionUID = 1L;
+
+    private int buff;
     private int healAmount;
-    private ArrayList<UnitModel> unitsToOverCharge;
+    private HashMap<Commander, ArrayList<Unit>> unitsOverCharged = new HashMap<Commander, ArrayList<Unit>>();
+    UnitTypeFilter damageBuff = null;
+    UnitModifier defenseBuff = null;
 
-    public OverchargeAbility(Commander commander, String abilityName, double abilityCost, int buff, int healAmt)
+    public OverchargeAbility(Tech tech, String abilityName, double abilityCost, int buff, int healAmt)
     {
-      super(commander, abilityName, abilityCost);
+      super(tech, abilityName, abilityCost);
 
-      // Create COModifiers that we can apply when needed.
-      damageBuff = new CODamageModifier(buff);
-      defenseBuff = new CODefenseModifier(buff);
+      this.buff = buff;
       healAmount = healAmt;
 
-      // Only mechanical/non-troop units get the firepower boost.
-      long exclude = UnitModel.TROOP;
-      unitsToOverCharge = commander.getAllModelsNot(exclude);
-
-      for( UnitModel m : unitsToOverCharge )
-        damageBuff.addApplicableUnitModel(m);
-
       AIFlags = PHASE_TURN_START | PHASE_TURN_END;
+
+      // Only mechanical/non-troop units get the firepower boost.
+      damageBuff = new UnitTypeFilter(new UnitDamageModifier(buff));
+      damageBuff.noneOf = UnitModel.TROOP;
+      defenseBuff = new UnitDefenseModifier(this.buff);
     }
 
-    private static final long serialVersionUID = 1L;
+    @Override
+    protected void enqueueUnitMods(MapMaster gameMap, ArrayList<UnitModifier> modList)
+    {
+      modList.add(damageBuff);
+      modList.add(defenseBuff);
+    }
 
     @Override
     protected void perform(MapMaster gameMap)
     {
-      // Bump up our power level.
-      myCommander.addCOModifier(damageBuff);
-      myCommander.addCOModifier(defenseBuff);
-
+      ArrayList<UnitModel> typesToOverCharge = getMechanicalModels(myCommander);
+      ArrayList<Unit> overCharged = new ArrayList<Unit>();
       // Overcharge
       for(Unit u: myCommander.units)
       {
-        if( unitsToOverCharge.contains(u.model) )
+        if( typesToOverCharge.contains(u.model) )
+        {
+          // Track units that aren't already overhealed
+          if( u.getHP() <= u.model.maxHP
+              && u.getHP() + healAmount > u.model.maxHP )
+            overCharged.add(u);
           u.alterHP(healAmount, true);
+        }
+      }
+      unitsOverCharged.put(myCommander, overCharged);
```

Comment:
Why bother storing the list of overcharged units instead of just trimming any >10 units when the power ends? Is it just to avoid iterating over the whole list again?

And why bother tracking them by commander? a) It is only ever applied to `myCommander` anyway, and b) who each unit belongs to has no impact on the revert logic.

I assume it has something to do with `Army` plans, but we don't have that yet; I'm pretty sure I've stated my opinion before on implementing logic to support things that don't exist yet.

Reply:
My intent is to cleanly interact with units that have >10 HP before the power starts, to handle other COs' units properly.

Tracking by commander was to account for the possibility of other COs activating this power for themselves, independent of the `Tech` instance they're stealing it from. Yeah, yeah, YAGNI and there are other design problems with it.
I'm not super attached to that idea.

I don't feel like YAGNI works quite so well in this instance, since this is possibly subtle bug territory. I'd rather change things now when I'm thinking about them than later when I'm maybe not thinking about them.

Reply:
> to handle other COs' units properly

This... doesn't handle other CO's units

> other COs activating this power for themselves

This is also not a thing that happens. Even in your alternate coop paradigms, nobody's activating somebody else's powers, right?

> subtle bug territory

How so? Seems more complex to manage unnecessary CO associations. Plus Tech just doesn't have another way to have units with >10 HP, so simply shaving anything higher than that is guaranteed to be correct. (unless you try real hard to contrive a scenario).

Reply:
> This... doesn't handle other CO's units

Overcharge seems like a perfectly reasonable thing to apply to all COs in an Army.
Sure, it doesn't do that *today*, but I'm really not interested in ripping out code I wrote to handle a case that I'm about to implement, just because I haven't yet implemented it.
> This is also not a thing that happens. Even in your alternate coop paradigms, nobody's activating somebody else's powers, right?

Yes, agreed. There is *some* vague possibility of that being in-scope eventually, but I imagine we can burn that bridge if and when we decide we want to cross it. (It occurs to me that probably the simple solution is to just clone the power for that weird case and not make everything more complex than it needs to be)
> How so? Seems more complex to manage unnecessary CO associations. Plus Tech just doesn't have another way to have units with >10 HP, so simply shaving anything higher than that is guaranteed to be correct. (unless you try real hard to contrive a scenario).

That it's more complex is strictly correct. That is the cost for handling more complex situations correctly.
You're also correct (today) that Tech can't get >10 HP units any other way, and that that behavior is correct today.
Technically speaking, we also don't have any COs *other* than Tech that can produce >10 HP units, and she always reverts on powers... so tags won't technically cause a problem with the COs we have either.
If we ever add a CO with >10 HP units outside of power turns, it'll suddenly become a problem.

Maybe the solution here is to formalize how >10 HP works - you aren't allowed to go above your max *ever* (so Tech adjusts the max accordingly), and damage is always calculated as if 10 HP is the max.
That's kinda how I expected it to work in the first place, but didn't object at the time because your implementation works and doesn't break anything today.

Reply:
> If we ever add a CO with >10 HP units outside of power turns, it'll suddenly become a problem

Only if Tech is attempting to revert her power on someone else's units, which shouldn't be a thing.

> Maybe the solution here is to formalize how >10 HP works - you aren't allowed to go above your max ever

There is a certain appeal to having a max that is actually a proper max. I probably didn't do it that way to start with because then you're into either mutating UnitModels or doing remodeling, and either way is extra complexity. Plus doing it this way offers an HP-based firepower buff which is also fun (but ofc we could just use a normal buff).

Reply:
> Only if Tech is attempting to revert her power on someone else's units, which shouldn't be a thing.

Which *should* be a thing, in like two PRs.
> There is a certain appeal to having a max that is actually a proper max.

Agreed.
> I probably didn't do it that way to start with because then you're into either mutating UnitModels or doing remodeling, and either way is extra complexity.

You had to change how `Unit` worked to accommodate that, though. Ideally, exceptions to rules that COs create should be located in the CO's file.
> Plus doing it this way offers an HP-based firepower buff which is also fun (but ofc we could just use a normal buff).

Again, we can define this however we like. We could have a "nominal max HP" that gets used for damage calculations, vs the "true max HP", which enforces how much HP you can have.
You're already using a normal buff - you explicitly wanted the multiplicative effect of overcharged HP alongside an attack buff. I'm not sure the numbers come out all that differently, but hey.

Reply:
> Which should be a thing, in like two PRs.

Right. The "one team, two armies" approach. Having like three different ways to handle allied COs is going to add a lot of subtle bug potential - I do hope you have a robust testing approach in the works to ensure it works as it should.

> not interested in ripping out code I wrote to handle a case that I'm about to implement, just because I haven't yet implemented it

This is understandable. On the other hand, I don't like to merge code changes into the release branch that aren't needed for anything, and may need to be changed again once we implement the proper use case. I won't make you change it back this time, but you really really need to get better at keeping your PRs focused.

Reply:
> Right. The "one team, two armies" approach. Having like three different ways to handle allied COs is going to add a lot of subtle bug potential - I do hope you have a robust testing approach in the works to ensure it works as it should.

*Team'd COs. Allies (as of this writing, anyway) aren't planned to interact meaningfully.
I imagine it would be sufficient to have a single set of operations on a game state defined with different tag behaviors, and observe whether it results in valid behavior in each individual case.
> This is understandable. On the other hand, I don't like to merge code changes into the release branch that aren't needed for anything

If there's a release branch, it's not `master` at the moment. But I get what you're saying.
> you really really need to get better at keeping your PRs focused.

Agreed.

Reply:
> If there's a release branch, it's not master

Why not? We shouldn't be merging stuff into master until it is known to be good.

Reply:
Because we've made releases, and none of them were from `master`?
To be fair, this is kind of irregular and ideally this shouldn't work this way.

Reply:
True enough; master should certainly be stable though.